syntax="proto3";

package descarteslabs.workflows;

message Typespec {
     oneof component {
          Primitive primitive = 1;
          string type = 2;
          Map map = 3;
          CompositeType composite = 4;
     }
}

message Primitive {
     oneof value {
          int32 int_  = 1;
          float float_ = 2;
          bool bool_ = 3;
          string string_ = 4;
     }
}

message MapFieldEntry {
     Typespec key = 1;
     Typespec val = 2;
}

message Map {
     repeated MapFieldEntry items = 1;
}

message CompositeType {
     string type = 1;
     repeated Typespec params = 2;
}
